### PR TITLE
Fixes #14373 - Response listener that avoids any data copies.

### LINKED ISSUE #14373: Response listener that avoids any data copies
**Jetty version(s)**

12.1.x

**Enhancement Description**

https://github.com/jetty/jetty.project/issues/12687 / https://github.com/jetty/jetty.project/pull/12782 introduced `RetainingResponseListener` with the aim to improve over `BufferingResponseListener`. The BRL accumulates data in its own buffers, while RRL retains incoming chunks, avoiding any data copies while request's response is being received.

However, both classes share one thing in common: the response content is available via `getContent()`, `getContentAsString()` or `getContentAsInputStream()`. Each of these methods calls `accumulator.takeByteArray()`, which copies data into a new `byte[]`.

There also exists `InputStreamResponseListener` which seems to do this. However, it's documented to require `InputStream` to be consumed _during_ request, otherwise the request may block.

**Proposal**: provide `RetainingResponseListener.getContentAsInputStream()` such that it's backed by individual chunks, and not `ByteArrayInputStream`.

cc @wendigo

### PR DESCRIPTION
**Pull request by sbordet**
* Implemented `AbstractResponseListener.getContentAsContentSource()` which can then be used for a more performant implementation of `getContentAsInputStream()` that does not perform data copy.
* Changed `HttpParser` to produce slices, rather than non-sliced read-only content ByteBuffers. In this way applications cannot peek before or after the content buffer. Removing read-only-ness would allow for testing, and it may be considered a feature, especially in the client where a content listener would be able to modify what the next content listener sees.
* Implemented `RetainableByteBuffer.DynamicCapacity.takeContentSource()` to pull out the content as a `Content.Source` without the data copy performed by `takeByteArray()`.

### CODE REVIEW

File: jetty-core/jetty-client/src/test/java/org/eclipse/jetty/client/RetainingResponseListenerTest.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+        // Modify the content so that we can check if there was a copy.
+        assertEquals(1, byteBuffers.size());
+        ByteBuffer byteBuffer = byteBuffers.get(0);
+        assertEquals(1, byteBuffer.remaining());
+        byte modified = 1;
+        byteBuffer.put(0, modified);
+
+        // Read from the input stream.
+        int read = inputStream.read();
```
Comment (lorban) [inline jetty-core/jetty-client/src/test/java/org/eclipse/jetty/client/RetainingResponseListenerTest.java]: Also check that the next few reads return -1?

File: jetty-core/jetty-client/src/test/java/org/eclipse/jetty/client/RetainingResponseListenerTest.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+import org.eclipse.jetty.server.Handler;
+import org.eclipse.jetty.server.Request;
+import org.eclipse.jetty.server.Response;
+import org.eclipse.jetty.util.Callback;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.ArgumentsSource;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class RetainingResponseListenerTest extends AbstractHttpClientServerTest
```
Comment (lorban) [inline jetty-core/jetty-client/src/test/java/org/eclipse/jetty/client/RetainingResponseListenerTest.java:34]: `AbstractHttpClientServerTest` does not check for buffer leaks. Now feels like the right time to add that logic.

File: jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/AbstractResponseListener.java [hunk 3]
```diff
@@ -189,7 +188,17 @@ public String getContentAsString(Charset charset)
...
     public InputStream getContentAsInputStream()
     {
-        return new ByteArrayInputStream(getContent());
+        return Content.Source.asInputStream(getContentAsContentSource());
+    }
+
+    /**
+     * @return the content as {@link Content.Source}
+     */
+    public Content.Source getContentAsContentSource()
```
Comment (lorban) [inline jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/AbstractResponseListener.java]: This method has the `get` prefix but acts as `take` (i.e.: just calling this method empties the accumulator) when the accumulator is a `DynamicCapacity` instance while it acts as a `get` when the accumulator is of any other type. This needs to be aligned: either both should consume the accumulator, or duplicate it.

File: jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/AbstractResponseListener.java [hunk 4]
```diff
@@ -189,7 +188,17 @@ public String getContentAsString(Charset charset)
...
+    {
+        if (accumulator instanceof RetainableByteBuffer.DynamicCapacity dynamic)
+            return dynamic.takeContentSource();
+        return Content.Source.from(accumulator.getByteBuffer());
```
Comment (lorban) [inline jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/AbstractResponseListener.java]: The returned content source contains a byte buffer whose lifecycle isn't controlled, so it may be put back into the pool at any time while the content source is in use.

You should use a `ChunksContentSource` and wrap the accumulator as a Chunk instead.

File: jetty-core/jetty-client/src/test/java/org/eclipse/jetty/client/RetainingResponseListenerTest.java [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    @ParameterizedTest
+    @ArgumentsSource(ScenarioProvider.class)
+    public void testContentIsNotCopiedWithInputStream(Scenario scenario) throws Exception
+    {
+        start(scenario, new Handler.Abstract()
+        {
+            @Override
+            public boolean handle(Request request, Response response, Callback callback)
+            {
+                response.write(true, ByteBuffer.allocate(1), callback);
```
Comment (lorban) [inline jetty-core/jetty-client/src/test/java/org/eclipse/jetty/client/RetainingResponseListenerTest.java:45]: I'd write at least a couple of buffers to test that the listener's content source can work with more than one.

**Review by lorban — CHANGES_REQUESTED** [review]

**Comment by wendigo** [comment]
We are getting exceptions in our tests:

```
java.util.concurrent.ExecutionException: java.util.NoSuchElementException
        at com.google.common.util.concurrent.AbstractFuture.getDoneValue(AbstractFuture.java:292)
        at com.google.common.util.concurrent.AbstractFutureState.blockingGet(AbstractFutureState.java:255)
        at com.google.common.util.concurrent.Platform.get(Platform.java:54)
        at com.google.common.util.concurrent.AbstractFuture.get(AbstractFuture.java:253)
        at io.airlift.http.client.AbstractHttpClientTest.testExecuteAsyncGetState(AbstractHttpClientTest.java:1018)
        at java.base/java.lang.reflect.Method.invoke(Method.java:565)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1604)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1604)
Caused by: java.util.NoSuchElementException
        at java.base/java.util.ArrayList$Itr.next(ArrayList.java:1053)
        at org.eclipse.jetty.io.content.ChunksContentSource.read(ChunksContentSource.java:85)
        at org.eclipse.jetty.io.content.ContentSourceInputStream.read(ContentSourceInputStream.java:84)
        at com.google.common.io.CountingInputStream.read(CountingInputStream.java:63)
        at java.base/java.io.InputStream.readNBytes(InputStream.java:411)
        at java.base/java.io.InputStream.readAllBytes(InputStream.java:348)
        at io.airlift.http.client.ResponseHandlerUtils.getResponseBytes(ResponseHandlerUtils.java:33)
        at io.airlift.http.client.StringResponseHandler.handle(StringResponseHandler.java:56)
        at io.airlift.http.client.StringResponseHandler.handle(StringResponseHandler.java:33)
        at io.airlift.http.client.jetty.JettyResponseFuture.processResponse(JettyResponseFuture.java:135)
        at io.airlift.http.client.jetty.JettyResponseFuture.completed(JettyResponseFuture.java:109)
        at io.airlift.http.client.jetty.JettyResponseListener.onComplete(JettyResponseListener.java:36)
        at org.eclipse.jetty.client.transport.ResponseListeners.notifyComplete(ResponseListeners.java:354)
        at org.eclipse.jetty.client.transport.ResponseListeners.lambda$addCompleteListener$0(ResponseListeners.java:339)
        at org.eclipse.jetty.client.transport.ResponseListeners.notifyComplete(ResponseListeners.java:354)
        at org.eclipse.jetty.client.transport.ResponseListeners.notifyComplete(ResponseListeners.java:346)
        at org.eclipse.jetty.client.transport.HttpReceiver.terminateResponse(HttpReceiver.java:454)
        at org.eclipse.jetty.client.transport.HttpReceiver.terminateResponse(HttpReceiver.java:436)
        at org.eclipse.jetty.client.transport.HttpReceiver.lambda$responseSuccess$0(HttpReceiver.java:394)
        at org.eclipse.jetty.util.thread.SerializedInvoker$Link.run(SerializedInvoker.java:274)
        at org.eclipse.jetty.util.thread.SerializedInvoker.run(SerializedInvoker.java:174)
        at org.eclipse.jetty.client.transport.HttpReceiver.responseHeaders(HttpReceiver.java:247)
        at org.eclipse.jetty.client.transport.internal.HttpReceiverOverHTTP.parse(HttpReceiverOverHTTP.java:335)
        at org.eclipse.jetty.client.transport.internal.HttpReceiverOverHTTP.parseAndFill(HttpReceiverOverHTTP.java:252)
        at org.eclipse.jetty.client.transport.internal.HttpReceiverOverHTTP.receive(HttpReceiverOverHTTP.java:85)
        at org.eclipse.jetty.client.transport.internal.HttpChannelOverHTTP.receive(HttpChannelOverHTTP.java:97)
        at org.eclipse.jetty.client.transport.internal.HttpConnectionOverHTTP.onFillable(HttpConnectionOverHTTP.java:267)
        at org.eclipse.jetty.client.transport.internal.HttpConnectionOverHTTP$FillableCallback.succeeded(HttpConnectionOverHTTP.java:458)
        at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105)
        at org.eclipse.jetty.io.ssl.SslConnection$SslEndPoint.onFillable(SslConnection.java:614)
        at org.eclipse.jetty.io.ssl.SslConnection.onFillable(SslConnection.java:415)
        at org.eclipse.jetty.io.ssl.SslConnection$2.succeeded(SslConnection.java:156)
        at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105)
        at org.eclipse.jetty.io.SelectableChannelEndPoint$1.run(SelectableChannelEndPoint.java:54)
        at org.eclipse.jetty.util.thread.MonitoredQueuedThreadPool$1.run(MonitoredQueuedThreadPool.java:73)
        at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:1009)
        at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.doRunJob(QueuedThreadPool.java:1239)
        at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1194)
        at java.base/java.lang.Thread.run(Thread.java:1474)
```

**Comment by wendigo** [comment]
Any news?

File: jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/AbstractResponseListener.java (hunk 4 above)
Comment (sbordet) [reply jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/AbstractResponseListener.java]: I am now using `getContent()` wrapped in a `ByteBuffer`.

File: jetty-core/jetty-client/src/test/java/org/eclipse/jetty/client/RetainingResponseListenerTest.java (hunk 5 above)
Comment (sbordet) [reply jetty-core/jetty-client/src/test/java/org/eclipse/jetty/client/RetainingResponseListenerTest.java:45]: I have written another test to test multiple buffers on the client (rather than on the server).

**Comment by sbordet** [comment]
@wendigo I updated the PR following the review.
Your failure is really strange, and it does not happen for us.
There is a lot of wrapping by `StringResponseHandler` and `CountingInputStream`, what are they doing?

**Comment by wendigo** [comment]
@sbordet these are parts of our airlift http client (https://github.com/airlift/airlift/tree/master/http-client)

**Comment by sbordet** [comment]
> @sbordet these are parts of our airlift http client (https://github.com/airlift/airlift/tree/master/http-client)

I know, which tests are failing?
As I said, it is strange that they fail, so either it's a corner case or airlift does something weird?

**Comment by wendigo** [comment]
@sbordet it worked so far so and it works in production so I doubt that it's wrong

**Comment by sbordet** [comment]
@wendigo ok, but what tests are failing?

**Comment by lorban** [comment]
@wendigo I updated this PR with a couple of fixes. We'd be glad if you could guide us on how to reproduce your failures so we can avoid round-trips between you testing and us fixing and also help us write a more robust test suite inside Jetty.

Thanks!

File: jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/AbstractResponseListener.java (hunk 4 above)
Comment (lorban) [reply jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/AbstractResponseListener.java]: That would still not give a `take` semantic. I've changed the `getContent()` call for `take()` instead.

**Comment by lorban** [comment]
@wendigo I've modified `io.airlift.http.client.jetty.JettyResponseListener` to extend from `RetainingResponseListener` instead of `BufferingResponseListener`, made airlift depend on jetty 12.1.7-SNAPSHOT and ran the whole airlift `http-client` test suite with `mvn verify`, and got a successful build.

Did you use anything else to try out this branch?

**Comment by wendigo** [comment]
@lorban that's what I did. I guess thet new version works now. I will retest soon

**Comment by wendigo** [comment]
Unfortunately tests are still failing for me locally: https://github.com/airlift/airlift/pull/1776

**Comment by wendigo** [comment]
```
❯ ./mvnw clean install -pl ':http-client'

[ERROR] io.airlift.http.client.jetty.TestJettyHttpClientHttpProxyAuthN.testExecuteAsyncGetState -- Time elapsed: 0.006 s <<< ERROR!
java.util.concurrent.ExecutionException: java.util.NoSuchElementException
        at com.google.common.util.concurrent.AbstractFuture.getDoneValue(AbstractFuture.java:292)
        at com.google.common.util.concurrent.AbstractFutureState.blockingGet(AbstractFutureState.java:255)
        at com.google.common.util.concurrent.Platform.get(Platform.java:54)
        at com.google.common.util.concurrent.AbstractFuture.get(AbstractFuture.java:253)
        at io.airlift.http.client.AbstractHttpClientTest.testExecuteAsyncGetState(AbstractHttpClientTest.java:1018)
        at java.base/java.lang.reflect.Method.invoke(Method.java:565)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1604)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1604)
Caused by: java.util.NoSuchElementException
        at java.base/java.util.ArrayList$Itr.next(ArrayList.java:1053)
        at org.eclipse.jetty.io.content.ChunksContentSource.read(ChunksContentSource.java:85)
        at org.eclipse.jetty.io.content.ContentSourceInputStream.read(ContentSourceInputStream.java:84)
        at com.google.common.io.CountingInputStream.read(CountingInputStream.java:63)
        at java.base/java.io.InputStream.readNBytes(InputStream.java:411)
        at java.base/java.io.InputStream.readAllBytes(InputStream.java:348)
        at io.airlift.http.client.ResponseHandlerUtils.getResponseBytes(ResponseHandlerUtils.java:33)
        at io.airlift.http.client.StringResponseHandler.handle(StringResponseHandler.java:56)
        at io.airlift.http.client.StringResponseHandler.handle(StringResponseHandler.java:33)
        at io.airlift.http.client.jetty.JettyResponseFuture.processResponse(JettyResponseFuture.java:135)
        at io.airlift.http.client.jetty.JettyResponseFuture.completed(JettyResponseFuture.java:109)
        at io.airlift.http.client.jetty.JettyResponseListener.onComplete(JettyResponseListener.java:35)
        at org.eclipse.jetty.client.transport.ResponseListeners.notifyComplete(ResponseListeners.java:354)
        at org.eclipse.jetty.client.transport.ResponseListeners.lambda$addCompleteListener$0(ResponseListeners.java:339)
        at org.eclipse.jetty.client.transport.ResponseListeners.notifyComplete(ResponseListeners.java:354)
        at org.eclipse.jetty.client.transport.ResponseListeners.notifyComplete(ResponseListeners.java:346)
        at org.eclipse.jetty.client.transport.HttpReceiver.terminateResponse(HttpReceiver.java:454)
        at org.eclipse.jetty.client.transport.HttpReceiver.terminateResponse(HttpReceiver.java:436)
        at org.eclipse.jetty.client.transport.HttpReceiver.lambda$responseSuccess$0(HttpReceiver.java:394)
        at org.eclipse.jetty.util.thread.SerializedInvoker$Link.run(SerializedInvoker.java:274)
        at org.eclipse.jetty.util.thread.SerializedInvoker.run(SerializedInvoker.java:174)
        at org.eclipse.jetty.client.transport.HttpReceiver.responseHeaders(HttpReceiver.java:247)
        at org.eclipse.jetty.client.transport.internal.HttpReceiverOverHTTP.parse(HttpReceiverOverHTTP.java:335)
        at org.eclipse.jetty.client.transport.internal.HttpReceiverOverHTTP.parseAndFill(HttpReceiverOverHTTP.java:252)
        at org.eclipse.jetty.client.transport.internal.HttpReceiverOverHTTP.receive(HttpReceiverOverHTTP.java:85)
        at org.eclipse.jetty.client.transport.internal.HttpChannelOverHTTP.receive(HttpChannelOverHTTP.java:97)
        at org.eclipse.jetty.client.transport.internal.HttpConnectionOverHTTP.onFillable(HttpConnectionOverHTTP.java:267)
        at org.eclipse.jetty.client.transport.internal.HttpConnectionOverHTTP$FillableCallback.succeeded(HttpConnectionOverHTTP.java:458)
        at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105)
        at org.eclipse.jetty.io.SelectableChannelEndPoint$1.run(SelectableChannelEndPoint.java:54)
        at org.eclipse.jetty.util.thread.MonitoredQueuedThreadPool$1.run(MonitoredQueuedThreadPool.java:73)
        at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:1009)
        at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.doRunJob(QueuedThreadPool.java:1239)
        at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1194)
        at java.base/java.lang.Thread.run(Thread.java:1474)
```

**Comment by wendigo** [comment]
@lorban I guess that you still used `getContent()` that copies all buffers to a single byte array and that's what we want to avoid

**Comment by wendigo** [comment]
Even if I do `byte[] bytes = getContentAsInputStream().readAllBytes();` in the `public void onComplete(Result result)` it will fail always with 
```
java.util.NoSuchElementException
	at java.base/java.util.ArrayList$Itr.next(ArrayList.java:1053)
	at org.eclipse.jetty.io.content.ChunksContentSource.read(ChunksContentSource.java:85)
	at org.eclipse.jetty.io.content.ContentSourceInputStream.read(ContentSourceInputStream.java:84)
	at java.base/java.io.InputStream.readNBytes(InputStream.java:411)
	at java.base/java.io.InputStream.readAllBytes(InputStream.java:348)
	at io.airlift.http.client.jetty.JettyResponseListener.onComplete(JettyResponseListener.java:40)
```

**Comment by wendigo** [comment]
This works:

```
    @Override
    public void onComplete(Result result)
    {
        if (result.isFailed()) {
            future.failed(result.getFailure());
        }
        else {
            try {
                byte[] bytes = getContent();
                future.completed(result.getResponse(), new ByteArrayInputStream(bytes));
            }
            catch (Exception e) {
                throw new RuntimeException(e);
            }
        }
    }
```

That doesn't:

```
    @Override
    public void onComplete(Result result)
    {
        if (result.isFailed()) {
            future.failed(result.getFailure());
        }
        else {
            try {
                byte[] bytes = getContentAsInputStream().readAllBytes();
                future.completed(result.getResponse(), new ByteArrayInputStream(bytes));
            }
            catch (Exception e) {
                throw new RuntimeException(e);
            }
        }
    }
```

**Comment by lorban** [comment]
@wendigo thanks for the explanation, I obviously missed the `getContent()` call in `onComplete()`.

This test of yours uncovered a couple of bugs. I just pushed the necessary fixes (so you can test them) and I'm now going to work on improving our test coverage to make sure those untested paths are now covered.

Please note that `RetainingResponseListener` is a bit special: if you call `getContent()` _before_ calling `getContentAsInputStream()` then the data is copied into a `byte[]` and the input stream will just work over that, but further `getContent*()` calls will always see the content. But if you call `getContentAsInputStream()` first, then the content is "stolen" by the input stream and further `getContent*()` calls will see an empty content.

I will add the above note to `RetainingResponseListener`'s javadoc.

**Comment by wendigo** [comment]
@lorban now it works :) I'm going to try that in the Trino itself in a sec

**Comment by wendigo** [comment]
@lorban I'm aware that these methods are mutually exclusive (fine for me)

**Comment by wendigo** [comment]
I've tested it with modified handlers that are doing streaming reads and it works fine :) We've halved our byte[] allocations basically.

**Comment by lorban** [comment]
@wendigo thanks for the feedback. Good to know that this helps.

@sbordet this PR is now fully done, and ready for a final review.

File: jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java [hunk 6]
```diff
@@ -2463,6 +2464,26 @@ protected void onCompleteFailure(Throwable x)
...
+                if (buffer instanceof Content.Chunk chunk)
+                    list.add(chunk);
+                else
+                    list.add(Content.Chunk.asChunk(buffer.getByteBuffer(), !iterator.hasNext(), buffer));
+            }
+            ChunksContentSource contentSource = new ChunksContentSource(list);
+            for (RetainableByteBuffer buffer : _buffers)
+                buffer.release();
+            _buffers.clear();
+            _aggregate = null;
```
Comment (sbordet) [inline jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java]: Replace with `clear()`.

File: jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java [hunk 7]
```diff
@@ -2463,6 +2464,26 @@ protected void onCompleteFailure(Throwable x)
...
+        {
+            List<Content.Chunk> list = new ArrayList<>();
+            Iterator<RetainableByteBuffer> iterator = _buffers.iterator();
+            while (iterator.hasNext())
+            {
+                RetainableByteBuffer buffer = iterator.next();
+                if (buffer instanceof Content.Chunk chunk)
+                    list.add(chunk);
+                else
+                    list.add(Content.Chunk.asChunk(buffer.getByteBuffer(), !iterator.hasNext(), buffer));
```
Comment (sbordet) [inline jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java]: `else if (buffer instanceof Dynamic dynamic) => list.addAll(dynamic._buffers)` -- something like that to "flatmap" the chunks.

File: jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/AbstractResponseListener.java [hunk 8]
```diff
@@ -189,7 +190,24 @@ public String getContentAsString(Charset charset)
...
-        return new ByteArrayInputStream(getContent());
+        if (content != null)
+            return new ByteArrayInputStream(content);
+        return Content.Source.asInputStream(getContentAsContentSource());
+    }
+
+    /**
+     * @return the content as {@link Content.Source}
+     */
+    public Content.Source getContentAsContentSource()
```
Comment (sbordet) [inline jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/AbstractResponseListener.java]: Rename to `takeContentAsInputStream()` to explicitate the "take" semantic.

Restore `getContentAsInputStream()` with non-take semantic.

Introduce `takeContentAsInputStream()` with take semantic.

File: jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java (hunk 6 above)
Comment (lorban) [reply jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java]: Done.

File: jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java (hunk 7 above)
Comment (lorban) [reply jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java]: Done.

File: jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/AbstractResponseListener.java (hunk 8 above)
Comment (lorban) [reply jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/AbstractResponseListener.java]: Done.

**Review by lorban — DISMISSED** [review]

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java [hunk 9]
```diff
@@ -811,7 +811,11 @@ private boolean reset(MetaData.Request request, MetaData.Response response, Byte
                 _content = content;
                 _lastContent = last;
                 _callback = callback;
-                _header = null;
+                if (_header != null)
+                {
+                    _header.release();
+                    _header = null;
+                }
```
Comment (sbordet) [inline jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java]: @lorban this should not be done here.
Release of `_header` should be done when the write completes of fails, not here.
Why was this necessary?

Comment (lorban) [reply jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java]: There was a leak of the `_header` RBB caused by the `SendCallback`'s reset method being called by the `onFailure()` method when the callback was failed. This was fixed by moving the failure of the callback to the `onCompleteFailure()` method.

**Review by lorban — APPROVED** [review]

**Comment by lorban** [comment]
@wendigo the feature got in and will be part of version 12.1.7.

Just a word of caution for you: to avoid breaking backward compatibility, the `getContentAsInputStream()` method kept its existing behavior: copying the content into a `byte[]` and returning an `InputStream` that reads over that. If you want to avoid the copy, you have to use the newly introduced `takeContentAsInputStream()` method.

**Comment by wendigo** [comment]
@lorban I've seen a discussion! Thanks
